numpy/lib/function_base.py

```diff
@@ -1631,7 +1631,7 @@ def trim_zeros(filt, trim='fb'):
         # Numpy 1.20.0, 2020-07-31
         warning = DeprecationWarning(
             "in the future trim_zeros will require a 1-D array as input "
-            "that is compatible with ndarray.astype(bool)"
+            "that supports elementwise comparisons with zero"
```

Comment:
We should move this deprecation to *after* the old implementation. Lets not worry too much about a possible changed exception type (although if you like that would be fine. But `None == 0` *always* failed, and should fail roughly identically. So we not should give a DeprecationWarning, when a correct error is given!

I do think we should add at least one tests for the `== 0` behaviour, and compare it to the old behaviour.

Reply:
The warning has been moved as of https://github.com/numpy/numpy/pull/17058/commits/f977f575d6263cac4703d809f57fe895415993ae.

Can you clarify a bit more about the `None == 0`? Are you talking about `filt=None` or `filt=[None, ...]` here? 
The former case would always raise an exception (and still does) while the latter would pass through `trim_zeros()` without any trimming (as `None != 0`). 
If we go ahead with this pull request then the behavior of the latter will also remain unchanged (though a warning will be issues due to the creation of an `object` array without explicitly specifying the `object` dtype).

Reply:
Yes, I was talking about the behaviour of the latter.  It creates an object array now, I do not think it gives a warning? But even if it does, I would be happy with that.

Reply:
>  It creates an object array now, I do not think it gives a warning?

You're right, I was confused with the deprecation of ragged arrays.

I think the current behavior of not issuing a warning should be fine, 
as `object` arrays will pass through `trim_zero()` just as they always have:
leading / trailing elements will be trimmed if they are equivalent to `0` and they'll be ignored otherwise.

In any case, I just added a few tests related to object arrays in https://github.com/numpy/numpy/pull/17058/commits/26734efdb1c8ee269bb97acc22b587027a5a3f88.